gdctools/gdc_mirror.py

```diff
@@ -182,14 +206,44 @@ def __mirror_file(self, file_d, proj_root, n, total, retries=3):
             if retry > retries:
                 # A partially downloaded file will interfere with subsequent
                 # mirrors
-                common.silent_rm(savepath)
-                logging.error("Error downloading file {0}, too many retries ({1})".format(savepath, retries))
+                common.silent_rm(partialsavepath)
+                logging.warning("Error downloading file {0}, too many retries ({1})".format(savepath, retries))
             else:
-                #Save md5 checksum on success
+                #success! First remove old md5, then rename file, then write new md5
+                #That way, if the process gets interrupted, future runs will be consider
+                # the file dirty unless each of these steps complete.
                 md5sum = file_d['md5sum']
                 md5path = savepath + ".md5"
+                if os.path.exists(md5path):
+                    os.remove(md5path)
+                os.rename(partialsavepath,savepath)
                 with open(md5path, 'w') as mf:
                     mf.write(md5sum + "  " + basename)
+                file_status = 'pass'
+
+        return file_status
+
+    @staticmethod
+    def __merge_mirror_metadata(old, new):
+        pass
+        file_item_dict = {}
+        for file_item in old:
+            key = file_item['file_id'] + file_item['md5sum']
+            file_item_dict[key] = file_item
+
+        if new is not None:
+            for file_item in new:
+                key = file_item['file_id'] + file_item['md5sum']
+                # TODO should there be a check whether the rest of the metadata matches?
+                file_item_dict[key] = file_item       
+
+        file_item_list = []    
+        file_item_keys = list(file_item_dict.keys())
+        file_item_keys.sort()
+        for key in file_item_keys:
+            file_item_list.append(file_item_dict[key])
+        return file_item_list
+
```

Comment:
```python
return [file_item_dict[key] for key in sorted(file_item_dict.keys())]
```

Reply:
How does this handle when there is a new version of a file? Should these keys really include the md5sum, or should they just be filenames?

Reply:
If there is a new version of the file, there will definitely be a new md5sum, and if the GDC does their job right there will also be a new file_id.  The sorting is just to keep the file easier for humans to work with; the key could also include filename, datatype, etc, in whatever order, in addition to the file_id and md5sum to force things to be unique.

If the same file is downloaded twice, the metadata from the second download will be recorded. 

If the file changes, both versions of the file will be kept, and the onus is on the loadfile generation to either select the right one or complain.   I haven't checked this case.

Reply:
The loadfile generator checks for duplicate samples, this will create duplicate aliquots, in which case one will simply be selected semi-randomly (the old method was to select the highest portion/plate value which in these cases will all be the same), unless the replicate filter calls that an error (I haven't checked). Perhaps the keys should be the aliquot IDs, that way the new will replace the old, which happens naturally in the normal mode of operations.

Reply:
The poorly named `lib.meta.tcga_id` function can be used to pull the proper ID. To handle clinical, the 'data_category' field could be used in conjunction with the ID to create unique keys.

Reply:
This appears to be an unresolved concern:  duplicates can either slip through or be chosen arbitrarily; but in master this concern does not exist;  this is grounds for not passing review.